### PR TITLE
ci: switch from service principal to managed identity for e2e test

### LINKED ISSUE #423: ci: switch from service principal to managed identity for e2e test

### PR DESCRIPTION
**Pull request by nilekhc**
Signed-off-by: Nilekh Chaudhari <1626598+nilekhc@users.noreply.github.com>

**Reason for Change**:

This PR switches to use managed identity for e2e azure tests

**Requirements**

- [x] squashed commits
- [x] included documentation
- [x] added unit tests and e2e tests (if applicable).

**Issue Fixed**:

fixes #423 
**Does this change contain code from or inspired by another project?**

- [ ] Yes
- [x] No

**If "Yes," did you notify that project's maintainers and provide attribution?**

**Special Notes for Reviewers**:

### CODE REVIEW

File: .pipelines/templates/role-assignment.yaml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+      echo "##vso[task.setvariable variable=ASSIGNEE_OBJECT_ID]${ASSIGNEE_OBJECT_ID}"
+
+      ROLE_ASSIGNMENT_IDS=""
+
+      az role assignment create --assignee-object-id "${ASSIGNEE_OBJECT_ID}" --assignee-principal-type "ServicePrincipal" --role "Virtual Machine Contributor" --scope "/subscriptions/$(SUBSCRIPTION_ID)/resourcegroups/$(CLUSTER_RESOURCE_GROUP)"
+      az role assignment create --assignee-object-id "${ASSIGNEE_OBJECT_ID}" --assignee-principal-type "ServicePrincipal" --role "Managed Identity Operator" --scope "/subscriptions/$(SUBSCRIPTION_ID)/resourcegroups/$(CLUSTER_RESOURCE_GROUP)"
+
+      # permission for Key Vault
+      if [[ -n "$(USER_MSI_RESOURCE_GROUP)" ]]; then
+        ID="$(az role assignment create --assignee-object-id "${ASSIGNEE_OBJECT_ID}" --assignee-principal-type "ServicePrincipal" --role "Reader" --scope "/subscriptions/$(SUBSCRIPTION_ID)/resourcegroups/$(USER_MSI_RESOURCE_GROUP)/providers/Microsoft.KeyVault/vaults/$(KEYVAULT_NAME)" --query id -otsv)"
```
Comment (aramase) [inline .pipelines/templates/role-assignment.yaml:14]: The CSI driver doesn't require this permission on the keyvault. Do we need this?

**Review by aramase — COMMENTED** [review]
Added 1 comment for role assignments.

1. Have we validated the nightly and load test pipeline with these changes?
2. I only see `e2e-azure-pr` in the PR status. Any ideas why the matrix is not being displayed?
3. What is the plan for moving the kind e2e pipeline also?

File: .pipelines/templates/role-assignment.yaml (hunk 1 above)
Comment (nilekhc) [reply .pipelines/templates/role-assignment.yaml]: yup. removed it. [build](https://dev.azure.com/AzureContainerUpstream/Secrets%20Store%20CSI%20Driver%20Provider%20Azure/_build/results?buildId=26667&view=results) is passing.

**Comment by nilekhc** [comment]
> 1. Have we validated the nightly and load test pipeline with these changes?

Yes. The build is [passing](https://dev.azure.com/AzureContainerUpstream/Secrets%20Store%20CSI%20Driver%20Provider%20Azure/_build/results?buildId=26940&view=results)

> 2. I only see `e2e-azure-pr` in the PR status. Any ideas why the matrix is not being displayed?

I think it might be because we still have triggers on for Azure org. We can verify this once we remove pipelines from there.
> 3. What is the plan for moving the kind e2e pipeline also?

Moved other [pipelines](https://dev.azure.com/AzureContainerUpstream/Secrets%20Store%20CSI%20Driver%20Provider%20Azure/_build) as well

File: .pipelines/templates/aks-setup.yaml [hunk 2]
```diff
@@ -87,6 +77,15 @@ steps:
         az aks nodepool add -g ${AZURE_CLUSTER_NAME} --cluster-name ${AZURE_CLUSTER_NAME} --os-type Windows --name win --node-count 1 ${EXTRA_ARGS:-} > /dev/null
       fi
 
+      # add gpu node pool
+      if [[ ${{ parameters.testWithGPU }} == True ]]; then
+        echo "adding gpu node pool"
+        az extension add --name aks-preview
+        az extension update --name aks-preview
+
```
Comment (aramase) [inline .pipelines/templates/aks-setup.yaml:85]: I see we're no longer setting the `AZURE_LOCATION` to `eastus2euap`. Just to confirm this is only available in westus other than eastus2euap?

File: .pipelines/templates/soak-test.yaml [hunk 3]
```diff
@@ -32,18 +28,20 @@ jobs:
...
 
             az aks get-credentials \
               --resource-group $(CLUSTER_RESOURCE_GROUP) \
               --name $(CLUSTER_CONFIG)
           displayName: "Set KUBECONFIG"
 
+        - ${{ if eq(clusterConfig, 'csi-secrets-store-soak-linux-aks') }}:
+          - script: |
+              echo "Installing helm..."
+              curl https://raw.githubusercontent.com/helm/helm/master/scripts/get-helm-3 | bash
```
Comment (aramase) [inline .pipelines/templates/soak-test.yaml:42]: Can we use the existing make target for installing helm?

File: .pipelines/templates/unit-test.yaml [hunk 4]
```diff
@@ -4,10 +4,9 @@ jobs:
   cancelTimeoutInMinutes: 5
   workspace:
     clean: all
+  variables:
+    - group: csi-secrets-store-e2e-kind
```
Comment (aramase) [inline .pipelines/templates/unit-test.yaml:8]: Why do we need this variable group for unit tests?

File: .pipelines/templates/soak-test.yaml (hunk 3 above)
Comment (nilekhc) [reply .pipelines/templates/soak-test.yaml:42]: updated

File: .pipelines/templates/unit-test.yaml (hunk 4 above)
Comment (nilekhc) [reply .pipelines/templates/unit-test.yaml]: We need `TRIVY_VERSION` for scan-images job.

Comment (aramase) [reply .pipelines/templates/unit-test.yaml:8]: That can be set as a pipeline variable

File: .pipelines/templates/soak-test.yaml (hunk 3 above)
Comment (nilekhc) [reply .pipelines/templates/soak-test.yaml]: Tests are [passing](https://dev.azure.com/AzureContainerUpstream/Secrets%20Store%20CSI%20Driver%20Provider%20Azure/_build/results?buildId=27052&view=results).

File: .pipelines/templates/unit-test.yaml (hunk 4 above)
Comment (nilekhc) [reply .pipelines/templates/unit-test.yaml]: makes sense. Updated.

**Review by aramase — APPROVED** [review]

**Pull request merged**